### PR TITLE
Add SwiftLint - Some basic cleanups + explicit self

### PR DESCRIPTION
**Pull request by freak4pc**
This PR started off to resolve using explicit `self.` invocation across the framework. SwiftLint introduced a new `analyze` command that analyzes the compiler flags and can automatically detect where `self.` is missing using the `explicit_self` rule.

I fixed all occurrences of missing `self.` but, along the way, added some useful SwiftLint rules and made some fixes around these.

I added each set of fixes (rule) in its own commit so I can "pull out" any change you don't necessarily like. 

Let me know your thoughts 🤓

### CODE REVIEW

File: RxCocoa/Common/DelegateProxyType.swift [hunk 1]
```diff
@@ -40,7 +40,7 @@ This is more or less how it works.
                   |                                                           
       +-----------v-------------------------------+                           
       |                                           |                           
-      | Delegate proxy : DelegateProxyType        +-----+---->  Observable<T1>
+      | Delegate proxy: DelegateProxyType        +-----+---->  Observable<T1>
```
Comment (kzaher) [inline RxCocoa/Common/DelegateProxyType.swift]: I think `+-----+---->  Observable<T1>` needs to be moved one space to the right.

File: RxCocoa/Common/RxTarget.swift [hunk 2]
```diff
@@ -10,7 +10,7 @@ import class Foundation.NSObject
 
 import RxSwift
 
-class RxTarget : NSObject
+class RxTarget: NSObject
                , Disposable {
```
Comment (kzaher) [inline RxCocoa/Common/RxTarget.swift]: This probably needs to be moved one space to the left.

File: RxCocoa/Foundation/NSObject+Rx.swift [hunk 3]
```diff
@@ -317,28 +317,28 @@ fileprivate final class DeallocObservable {
...
     var target: AnyObject { get }
     var keyPath: String { get }
     var retainTarget: Bool { get }
     var options: KeyValueObservingOptions { get }
 }
 
 fileprivate final class KVOObserver
-    : _RXKVOObserver
+   : _RXKVOObserver
     , Disposable {
```
Comment (kzaher) [inline RxCocoa/Foundation/NSObject+Rx.swift]: Something is wrong with alignment.

File: RxCocoa/Foundation/NSObject+Rx.swift [hunk 4]
```diff
@@ -362,7 +362,7 @@ fileprivate final class KVOObserver
 }
 
 fileprivate final class KVOObservable<Element>
-    : ObservableType
+   : ObservableType
     , KVOObservableProtocol {
```
Comment (kzaher) [inline RxCocoa/Foundation/NSObject+Rx.swift]: Alignment is wrong here.

File: RxCocoa/Foundation/NSObject+Rx.swift [hunk 5]
```diff
@@ -495,7 +495,7 @@ fileprivate extension KeyValueObservingOptions {
 
                 let nextElementsObservable = keyPathSections.count == 1
                     ? Observable.just(nextTarget)
-                    : observeWeaklyKeyPathFor(nextObject!, keyPathSections: remainingPaths, options: options)
+                   : observeWeaklyKeyPathFor(nextObject!, keyPathSections: remainingPaths, options: options)
```
Comment (kzaher) [inline RxCocoa/Foundation/NSObject+Rx.swift]: Wrong alignment.

File: RxCocoa/Foundation/URLSession+Rx.swift [hunk 6]
```diff
@@ -28,7 +28,7 @@ import RxSwift
 
 /// RxCocoa URL errors.
 public enum RxCocoaURLError
-    : Swift.Error {
+   : Swift.Error {
```
Comment (kzaher) [inline RxCocoa/Foundation/URLSession+Rx.swift]: Alignment ....

File: RxCocoa/Foundation/URLSession+Rx.swift [hunk 7]
```diff
@@ -40,7 +40,7 @@ public enum RxCocoaURLError
 }
 
 extension RxCocoaURLError
-    : CustomDebugStringConvertible {
+   : CustomDebugStringConvertible {
```
Comment (kzaher) [inline RxCocoa/Foundation/URLSession+Rx.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxCollectionViewReactiveArrayDataSource.swift [hunk 8]
```diff
@@ -13,7 +13,7 @@ import RxSwift
 
 // objc monkey business
 class _RxCollectionViewReactiveArrayDataSource
-    : NSObject
+   : NSObject
     , UICollectionViewDataSource {
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxCollectionViewReactiveArrayDataSource.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxCollectionViewReactiveArrayDataSource.swift [hunk 9]
```diff
@@ -39,7 +39,7 @@ class _RxCollectionViewReactiveArrayDataSource
 }
 
 class RxCollectionViewReactiveArrayDataSourceSequenceWrapper<S: Sequence>
-    : RxCollectionViewReactiveArrayDataSource<S.Iterator.Element>
+   : RxCollectionViewReactiveArrayDataSource<S.Iterator.Element>
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxCollectionViewReactiveArrayDataSource.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxCollectionViewReactiveArrayDataSource.swift [hunk 10]
```diff
@@ -58,12 +58,12 @@ class RxCollectionViewReactiveArrayDataSourceSequenceWrapper<S: Sequence>
 
 // Please take a look at `DelegateProxyType.swift`
 class RxCollectionViewReactiveArrayDataSource<Element>
-    : _RxCollectionViewReactiveArrayDataSource
+   : _RxCollectionViewReactiveArrayDataSource
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxCollectionViewReactiveArrayDataSource.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxPickerViewAdapter.swift [hunk 11]
```diff
@@ -31,7 +31,7 @@ class RxPickerViewArrayDataSource<T>: NSObject, UIPickerViewDataSource, Sectione
 }
 
 class RxPickerViewSequenceDataSource<S: Sequence>
-    : RxPickerViewArrayDataSource<S.Iterator.Element>
+   : RxPickerViewArrayDataSource<S.Iterator.Element>
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxPickerViewAdapter.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxPickerViewAdapter.swift [hunk 12]
```diff
@@ -45,7 +45,7 @@ class RxPickerViewSequenceDataSource<S: Sequence>
 }
 
 final class RxStringPickerViewAdapter<S: Sequence>
-    : RxPickerViewSequenceDataSource<S>
+   : RxPickerViewSequenceDataSource<S>
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxPickerViewAdapter.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxTableViewReactiveArrayDataSource.swift [hunk 13]
```diff
@@ -13,7 +13,7 @@ import RxSwift
 
 // objc monkey business
 class _RxTableViewReactiveArrayDataSource
-    : NSObject
+   : NSObject
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxTableViewReactiveArrayDataSource.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxTableViewReactiveArrayDataSource.swift [hunk 14]
```diff
@@ -39,7 +39,7 @@ class _RxTableViewReactiveArrayDataSource
 
 
 class RxTableViewReactiveArrayDataSourceSequenceWrapper<S: Sequence>
-    : RxTableViewReactiveArrayDataSource<S.Iterator.Element>
+   : RxTableViewReactiveArrayDataSource<S.Iterator.Element>
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxTableViewReactiveArrayDataSource.swift]: Alignment ...

File: RxCocoa/iOS/DataSources/RxTableViewReactiveArrayDataSource.swift [hunk 15]
```diff
@@ -57,11 +57,11 @@ class RxTableViewReactiveArrayDataSourceSequenceWrapper<S: Sequence>
 
 // Please take a look at `DelegateProxyType.swift`
 class RxTableViewReactiveArrayDataSource<Element>
-    : _RxTableViewReactiveArrayDataSource
+   : _RxTableViewReactiveArrayDataSource
```
Comment (kzaher) [inline RxCocoa/iOS/DataSources/RxTableViewReactiveArrayDataSource.swift]: Alignment ...

File: RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift [hunk 16]
```diff
@@ -21,7 +21,7 @@ fileprivate let collectionViewPrefetchDataSourceNotSet = CollectionViewPrefetchD
 
 @available(iOS 10.0, tvOS 10.0, *)
 fileprivate final class CollectionViewPrefetchDataSourceNotSet
-    : NSObject
+   : NSObject
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift]: Alignment ...

File: RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift [hunk 17]
```diff
@@ -30,7 +30,7 @@ fileprivate final class CollectionViewPrefetchDataSourceNotSet
 
 @available(iOS 10.0, tvOS 10.0, *)
 open class RxCollectionViewDataSourcePrefetchingProxy
-    : DelegateProxy<UICollectionView, UICollectionViewDataSourcePrefetching>
+   : DelegateProxy<UICollectionView, UICollectionViewDataSourcePrefetching>
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift]: Alignment ...

File: RxCocoa/iOS/Proxies/RxCollectionViewDataSourceProxy.swift [hunk 18]
```diff
@@ -18,7 +18,7 @@ extension UICollectionView: HasDataSource {
 fileprivate let collectionViewDataSourceNotSet = CollectionViewDataSourceNotSet()
 
 fileprivate final class CollectionViewDataSourceNotSet
-    : NSObject
+   : NSObject
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxCollectionViewDataSourceProxy.swift]: Alignment ...

File: RxCocoa/iOS/Proxies/RxCollectionViewDataSourceProxy.swift [hunk 19]
```diff
@@ -34,7 +34,7 @@ fileprivate final class CollectionViewDataSourceNotSet
 
 /// For more information take a look at `DelegateProxyType`.
 open class RxCollectionViewDataSourceProxy
-    : DelegateProxy<UICollectionView, UICollectionViewDataSource>
+   : DelegateProxy<UICollectionView, UICollectionViewDataSource>
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxCollectionViewDataSourceProxy.swift]: Alignment ...

File: RxCocoa/iOS/Proxies/RxCollectionViewDelegateProxy.swift [hunk 20]
```diff
@@ -13,7 +13,7 @@ import RxSwift
 
 /// For more information take a look at `DelegateProxyType`.
 open class RxCollectionViewDelegateProxy
-    : RxScrollViewDelegateProxy
+   : RxScrollViewDelegateProxy
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxCollectionViewDelegateProxy.swift]: Alignment ...

File: RxCocoa/iOS/Proxies/RxNavigationControllerDelegateProxy.swift [hunk 21]
```diff
@@ -17,9 +17,9 @@
 
     /// For more information take a look at `DelegateProxyType`.
     open class RxNavigationControllerDelegateProxy
-        : DelegateProxy<UINavigationController, UINavigationControllerDelegate>
-        , DelegateProxyType 
-        , UINavigationControllerDelegate {
+       : DelegateProxy<UINavigationController, UINavigationControllerDelegate>,
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxNavigationControllerDelegateProxy.swift]: Alignment ...

**Comment by freak4pc** [comment]
Thanks @kzaher - What are your alignment rules in general for these cases? do you just align to the world "class" ? Because its not aligned to anything else, so I'm confused about your style rules here.

**Comment by freak4pc** [comment]
Also don't worry about writing down all spots. I'll go through them once I get your feedback on your style guide.

**Comment by kzaher** [comment]
> Thanks @kzaher - What are your alignment rules in general for these cases? do you just align to the word "class" ? Because its not aligned to anything else, so I'm confused about your style rules here.

Well I guess I've used multiples of 4 spaces for indentation.

```swift
    : XXX
    , YYY
    , ZZZ
```

> Also don't worry about writing down all spots. I'll go through them once I get your feedback on your style guide.

I'm not sure writing `alignment` hundreds of times if the best use of my time. It is pretty easy to spot misalignment like the ones I've noted. The diff is `+4,173 −3,700`. By all metrics this is one huge diff.

**Comment by freak4pc** [comment]
@kzaher I believe the alignment issues should be fixed now.  

1. Please note that the style of putting the first dependency on a separate line, and then additional ones in their own lines is a bit uncommon in the Swift world, which is why I was mistaken in handling it. It should be resolved. I didn't request in any way that you would review spacings this way. 

2. The changes in this PR mostly reflect the same "style things" - I'm aware this is a big diff, but there's not much to do to make it smaller as its just a lot of little things. As I mentioned, every "set" of fixes is in its own commit, so if there is some rule you dislike for this project, just let me know and I'll drop it out of this PR. 

Thanks for your time again, and let me know if you have any additional questions.

File: RxSwift/Observables/Catch.swift [hunk 22]
```diff
@@ -169,7 +169,7 @@ final private class Catch<Element>: Producer<Element> {
 // catch enumerable
 
 final private class CatchSequenceSink<S: Sequence, O: ObserverType>
-   : TailRecursiveSink<S, O>
+    : TailRecursiveSink<S, O>
```
Comment (freak4pc) [inline RxSwift/Observables/Catch.swift]: Just FYI this was misaligned before my PR (with a few others), so I've aligned those back.

File: RxExample/RxExample-iOSTests/Mocks/MockWireframe.swift [hunk 23]
```diff
@@ -9,12 +9,12 @@
...
-class MockWireframe : Wireframe {
-    let _openURL: (URL) -> ()
+class MockWireframe: Wireframe {
+    let _openURL: (URL) -> Void
     let _promptFor: (String, Any, [Any]) -> Observable<Any>
 
-    init(openURL: @escaping (URL) -> () = notImplementedSync(),
-        promptFor: @escaping (String, Any, [Any]) -> Observable<Any> = notImplemented()) {
+    init(openURL: @escaping (URL) -> Void = notImplementedSync(),
+         promptFor: @escaping (String, Any, [Any]) -> Observable<Any> = notImplemented()) {
```
Comment (kzaher) [inline RxExample/RxExample-iOSTests/Mocks/MockWireframe.swift]: This looks like it's indented wrong.

File: Tests/RxSwiftTests/Observable+CombineLatestTests+arity.tt [hunk 24]
```diff
@@ -119,7 +119,7 @@ extension ObservableCombineLatestTest {
         let factories: [(<%= (Array(0..<i).map { _ in "TestableObservable<Int>" }).joined(separator: ", ") %>) -> Observable<Int>] =
             [
                 { <%= (Array(0..<i).map { "e\($0)" }).joined(separator: ", ") %> in
-                    Observable.combineLatest(<%= (Array(0..<i).map { "e\($0)" }).joined(separator: ", ") %>) { (<%= (Array(0..<i).map { _ in "_" }).joined(separator: ", ") %>) -> Int in
+                    Observable.combineLatest(<%= (Array(0..<i).map { "e\($0)" }).joined(separator: ", ") %>) { (    <%= (Array(0..<i).map { _ in "_" }).joined(separator: ", ") %>) -> Int in
```
Comment (kzaher) [inline Tests/RxSwiftTests/Observable+CombineLatestTests+arity.tt]: Unsure why was this space added?

File: Tests/RxSwiftTests/Observable+ZipTests+arity.tt [hunk 25]
```diff
@@ -33,7 +33,7 @@ extension ObservableZipTest {
```
Comment (kzaher) [inline Tests/RxSwiftTests/Observable+ZipTests+arity.tt]: This is the source of truth for `Tests/RxSwiftTests/Observable+ZipTests+arity.swift`. This should probably be fixed also.

File: Tests/RxSwiftTests/Observable+CombineLatestTests+arity.tt [hunk 26]
```diff
@@ -119,7 +119,7 @@ extension ObservableCombineLatestTest {
         let factories: [(<%= (Array(0..<i).map { _ in "TestableObservable<Int>" }).joined(separator: ", ") %>) -> Observable<Int>] =
```
Comment (kzaher) [inline Tests/RxSwiftTests/Observable+CombineLatestTests+arity.tt]: This is the source of truth for `Tests/RxSwiftTests/Observable+CombineLatestTests+arity.swift`. This should probably be fixed also.

File: RxExample/RxExample-iOSTests/Mocks/MockWireframe.swift (hunk 23 above)
Comment (kzaher) [reply RxExample/RxExample-iOSTests/Mocks/MockWireframe.swift]: Oh, no, this is ok.

**Comment by freak4pc** [comment]
> We should probably add style checks to travis to enforce the style.
This should be done on a separate PR, we need to define what kind of checks we want to add, etc. 

For the other two points, I'll take a look here. Thanks !

File: Tests/RxSwiftTests/Observable+CombineLatestTests+arity.tt (hunk 24 above)
Comment (freak4pc) [reply Tests/RxSwiftTests/Observable+CombineLatestTests+arity.tt]: Looks like a mistake

**Comment by freak4pc** [comment]
@kzaher I think this covers the notes you mentioned:

- Softlinks reverted back
- Fixed `.tt` files and ran Preprocessor script to make sure it generates the right content. 

I'll address adding lint to CI in a separate PR. 

Let me know if you see anything else.

**Comment by freak4pc** [comment]
Let me know if you'd like me to change anything - I wouldn't want this to diverge too much with develop and there are many small changes here (more so if you'll merge RxAtomic updates before merging this). Thanks! 🙏

**Comment by kzaher** [comment]
Hi @freak4pc, I think that your changes revert https://github.com/ReactiveX/RxSwift/commit/bc3021591346940988c6fefe9c9dcc6ae372020e

**Comment by kzaher** [comment]
I think that tt file needs to be fixed probably.

**Comment by freak4pc** [comment]
I think he didn’t update the tt in that PR. Sure, I’m on it

**Comment by kzaher** [comment]
Let's not do something like this ever again. Any big changes will need to be split into separate PRs.

I can't review these kinds of changes in future.

**Comment by freak4pc** [comment]
Sure. Its definitely hard to split when it’s all related to the same thing (lint), but will take into consideration in the future. 

Thanks !

**Pull request merged**